src/target/riscv/riscv-013.c

```diff
@@ -1688,6 +1733,23 @@ static int examine(struct target *target)
 					info->index);
 			return ERROR_FAIL;
 		}
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_DBGRQ;
+	}
+
+	/* FIXME: This is needed since register_read_direct relies on target->state
+	 * to work correctly, so, if target->state does not represent current state
+	 * of target, e.g. if a target is halted, but target->state is
+	 * TARGET_UNKNOWN, it can fail early, (e.g. accessing registers via program
+	 * buffer can not be done atomically on a running hart becuse mstatus can't
+	 * be prepared for a register access and then restored)
+	 */
+	const enum target_state saved_tgt_state = target->state;
+	const enum target_debug_reason saved_dbg_reason = target->debug_reason;
+
+	if (target->state != TARGET_HALTED) {
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_DBGRQ;
```

Comment:
IIUC, at this point the target is guaranteed to be halted. So is it possible to simply change the `target->state` directly (as shonw below)?

I don't think it is necessary to store the previous state of the target.

```suggestion
	}

	/* Record that the target is now halted. This is needed for register 
	   access functions that work with progbuf to operate. */
	target->state = TARGET_HALTED;
	target->debug_reason = DBG_REASON_DBGRQ;
```

Reply:
Addressed [below](https://github.com/riscv/riscv-openocd/pull/842#discussion_r1182698676)